src/components/ProgramLoader/ProgramFileUpload.tsx

```diff
@@ -2,12 +2,58 @@ import { ExternalLink } from "lucide-react";
 import { Input } from "../ui/input";
 import { ProgramUploadFileOutput } from "./types";
 import { mapUploadFileInputToOutput } from "./utils";
-import { decodeStandardProgram } from "@typeberry/pvm-debugger-adapter";
-import { RegistersArray } from "@/types/pvm.ts";
+import { MemorySegment, decodeStandardProgram } from "@typeberry/pvm-debugger-adapter";
+import { MemoryChunkItem, PageMapItem, RegistersArray } from "@/types/pvm.ts";
 import { SafeParseReturnType, z } from "zod";
 import { useAppSelector } from "@/store/hooks";
 import { selectInitialState } from "@/store/debugger/debuggerSlice";
 
+const PAGE_SHIFT = 12; // log_2(4096)
+const PAGE_SIZE = 1 << PAGE_SHIFT;
+function pageAlign(v: number) {
+  return (((v + PAGE_SIZE - 1) >> PAGE_SHIFT) << PAGE_SHIFT) >>> 0;
+}
+
+function asChunks(mem: MemorySegment[]): MemoryChunkItem[] {
+  const items = [];
+  for (const segment of mem) {
+    if (segment.data === null) {
+      continue;
+    }
+    let data = segment.data;
+    let address = segment.start;
+    while (data.length > 0) {
+      const lenForPage = address % PAGE_SIZE === 0 ? PAGE_SIZE : pageAlign(address) - address;
+      const contents = Array.from<number>(data.subarray(0, Math.min(data.length, lenForPage)));
+      items.push({
+        address,
+        // TODO [ToDr] TEMPORARY until sikora fixes the boundaries.
+        contents: contents.slice(0, Math.min(PAGE_SIZE - 1, contents.length)),
+      });
+
+      // move data & address
+      data = data.subarray(contents.length);
+      address += contents.length;
+    }
+  }
+  return items;
+}
+function asPageMap(mem: MemorySegment[], isWriteable: boolean): PageMapItem[] {
+  const items = [];
+  for (const segment of mem) {
+    const pageStart = segment.start % PAGE_SIZE === 0 ? segment.start : pageAlign(segment.start - PAGE_SIZE + 1);
```

Comment:
"else" looks incorrect here. I think it should be  `pageAlign(segment.start - PAGE_SIZE) + 1;`.

Reply:
I think it's correct. We basically subtract `PAGE_SIZE - 1` from starting address to make sure that we have an index on the previous page and then `pageAlign` upwards to the full page.

Your proposition will not work, since you aling to the start of the page but then always add `+1` (so we are always 1 index in the page already.

That said, I rewrote that part of the code to make it simpler, i.e. not using `pageAlign` but rather doing `const pageStart = segment.start - (segment.start % PAGE_SIZE)`;

Reply:
> I think it's correct. We basically subtract PAGE_SIZE - 1 from starting address to make sure that we have an index on the previous page and then pageAlign upwards to the full page.

ah okay, now I get it

> That said, I rewrote that part of the code to make it simpler, i.e. not using pageAlign but rather doing const pageStart = segment.start - (segment.start % PAGE_SIZE);

yeah this is much better